ceno_host/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+// Everything else in here would be hidden.
+
+use itertools::izip;
+use rkyv::{
+    Portable, Serialize,
+    api::high::{HighSerializer, HighValidator},
+    bytecheck::CheckBytes,
+    rancor::Error,
+    ser::allocator::ArenaHandle,
+    to_bytes,
+    util::AlignedVec,
+};
+
+#[derive(Default)]
+pub struct CenoStdin {
+    pub items: Vec<AlignedVec>,
+}
+
+pub struct SerialisedCenoStdin(pub AlignedVec);
+
+impl CenoStdin {
+    pub fn write_slice(&mut self, bytes: AlignedVec) {
+        self.items.push(bytes);
+    }
+
+    pub fn write(
+        &mut self,
+        item: &impl for<'a> Serialize<HighSerializer<AlignedVec, ArenaHandle<'a>, Error>>,
+    ) -> Result<(), Error> {
+        let bytes = to_bytes::<Error>(item)?;
+        self.write_slice(bytes);
+        Ok(())
+    }
+
+    pub fn finalise(&self) -> SerialisedCenoStdin {
+        // TODO: perhaps don't hardcode 16 here.
+        // It's from rkyv's format, so we can probably take it from there somehow?
+        // TODO: clean this up.
+        let initial_offset = (size_of::<u32>() * self.items.len()).next_multiple_of(16);
+        println!("offset: {}", initial_offset);
+        let offsets: Vec<u32> = self
+            .items
+            .iter()
+            .scan(initial_offset, |acc, bytes| {
+                let output = (*acc + bytes.len()) as u32;
+                print!("len: {}\t", bytes.len());
+                *acc += bytes.len().next_multiple_of(16);
+                println!("acc: {}", *acc);
+                Some(output)
+            })
+            .collect();
+        let offsets_u8: Vec<u8> = offsets.iter().copied().flat_map(u32::to_le_bytes).collect();
+        let mut buf: AlignedVec = AlignedVec::new();
+        buf.extend_from_slice(&offsets_u8);
+        println!("buf.len() after offsets: {}", buf.len());
+        buf.extend_from_slice(&vec![0; buf.len().next_multiple_of(16) - buf.len()]);
+        println!("buf.len() after offset padding: {}", buf.len());
+        for (offset, item) in izip!(offsets, &self.items) {
+            buf.extend_from_slice(item);
+            buf.extend_from_slice(&vec![0; buf.len().next_multiple_of(16) - buf.len()]);
+            assert_eq!(buf.len(), offset.next_multiple_of(16) as usize);
+        }
+        SerialisedCenoStdin(buf)
+    }
+}
+
+pub struct SerialisedCenoStdinIter<'a> {
+    buf: &'a SerialisedCenoStdin,
+    next: usize,
+}
+
+impl<'b> SerialisedCenoStdinIter<'b> {
+    pub fn read<'a, T>(&'a mut self) -> &'b T
+    where
+        T: Portable + for<'c> CheckBytes<HighValidator<'c, Error>>,
+    {
+        rkyv::access::<T, Error>(self.read_slice()).unwrap()
+    }
+
+    pub fn read_slice<'a>(&'a mut self) -> &'b [u8] {
+        self.next().unwrap()
+    }
+}
+
+impl<'a> Iterator for SerialisedCenoStdinIter<'a> {
+    type Item = &'a [u8];
+    fn next(&mut self) -> Option<Self::Item> {
+        let len = u32::from_le_bytes(
+            self.buf.0[self.next..][..size_of::<u32>()]
```

Comment:
@kunxian-xia So we actually only store 'pointers' to the end of the slices at the start of our hints memory region.

That's because rkyv doesn't mind, if we give it a slice that has extra garbage at the beginning, so without loss of generality, all the slices here start at the beginning of the hints memory region.  They are overlapping.

We can get away with this, because it's all read-only.  If you want to go read write, you'll want non-overlapping slices, which is also doable at only a slight extra overhead.

Reply:
Crucially, I have also not put in any information to delineate the end of the first section that stores the slice lengths / end points: a well-intentioned prover will provide all the entries that the guest program will want to read anyway, and a malicious prover could just give you misleading information.

If you accidentally read beyond the intended end, your deserialisation will (likely) fail.

However, I'm open to adding an explicit end, if you think that's useful for eg debugging, even if it makes no difference for security.  Just let me know, what you think, please.